### PR TITLE
Installer work in a mergeable chunk

### PR DESCRIPTION
**Pull request by camronlevanger**

### CODE REVIEW

File: packages/toolbox/src/swarm.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  const keyObject = keythereum.dump(
+    password,
+    dk.privateKey,
+    dk.salt,
+    dk.iv,
+    options,
+  )
+  await fs.ensureDir('keystore')
```
Comment (PaulLeCam) [inline packages/toolbox/src/swarm.js]: Might be safer to also delete this folder after the key is copied.

File: packages/toolbox/src/swarm.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  const path = await keythereum.exportToFile(keyObject)
+  await fs.ensureDir(datadir)
```
Comment (PaulLeCam) [inline packages/toolbox/src/swarm.js]: Not a big issue but every `await` makes the calls happen in sequence, while some actions could happen in parallel, such as the creation of these folders that could be changed to:
```js
await Promise.all([fs.ensureDir('keystore'), fs.ensureDir(datadir)])
```

File: packages/toolbox/src/swarm.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  return keyObject.address
+}
+
+export const importKeyStore = async (address: string): Promise<KeyObject> => {
+  return keythereum.importFromFile(address, datadir)
+}
+
+export const getPrivateKey = async (
+  address: string,
+  encoding: string,
```
Comment (PaulLeCam) [inline packages/toolbox/src/swarm.js]: Maybe we can make this second argument optional by setting a default value, `encoding?: string = 'hex'`

File: packages/toolbox/src/swarm.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  const keystores = await fs.readdir(`${datadir}/keystore`)
+  // eslint-disable-next-line no-console
+  console.log('Existing keystore files:')
+  // eslint-disable-next-line no-console
+  console.log(keystores)
+  if (keystores.length >= 1) {
+    keyObjects = []
+    for (i = 0; i < keystores.length; ++i) {
+      const address = keystores[i].substr(keystores[i].length - 40)
+      const keyObject = await importKeyStore(address)
```
Comment (PaulLeCam) [inline packages/toolbox/src/swarm.js]: Having an `await` in a loop will make everything happen in sequence, in order to run the `importKeyStore()` calls in parallel this code could be replaced by:
```js
const keyObjects = await Promise.all(keystores.map(async key => {
  return await importKeyStore(key.slice(40))
}))
```

File: packages/toolbox/src/swarm.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  return keyBuffer.toString(encoding)
+}
+
+export const listKeyStores = async (): Promise<KeyObject[]> => {
+  fs.ensureDir(datadir)
+  let i = 0
+  let keyObjects
+  const keystores = await fs.readdir(`${datadir}/keystore`)
+  // eslint-disable-next-line no-console
+  console.log('Existing keystore files:')
```
Comment (PaulLeCam) [inline packages/toolbox/src/swarm.js]: I'd rather not have these logs in the library, if the output is relevant we can return it to the caller.

File: packages/daemon/package.json [hunk 6]
```diff
@@ -57,9 +57,10 @@
     "ethereumjs-tx": "^1.3.7",
     "ethereumjs-util": "^6.0.0",
     "ethereumjs-wallet": "^0.6.2",
-    "fs-extra": "^7.0.0",
+    "fs-extra": "^7.0.1",
     "graphql": "^0.13.2",
     "graphql-relay": "^0.5.5",
+    "keythereum": "^1.0.4",
```
Comment (PaulLeCam) [inline packages/daemon/package.json]: Do we need to add this dependency to the daemon please? There doesn't seem to be any code change in the daemon using it.

File: package.json [hunk 7]
```diff
@@ -1,8 +1,12 @@
 {
   "name": "js-mainframe",
+  "productName": "MainframeOS",
```
Comment (PaulLeCam) [inline package.json]: Could you revert these changes please? We shouldn't add `electron` as a dependency and build setup here, better use the launcher package for this.

File: packages/toolbox/src/swarm.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      salt: string,
+    },
+  },
+  id: string,
+  version: number,
+}
+
+const homedir = os.homedir()
+const password = 'mainframe'
+const datadir = `${homedir}/.ethereum`
```
Comment (howleysv) [inline packages/toolbox/src/swarm.js]: Might be better to have under our own namespace. eg `.mainframe-os/services/swarm` or something?

Is this format portable to windows? `/` vs `\`

File: packages/toolbox/src/swarm.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+  const keyObject = keythereum.dump(
+    password,
+    dk.privateKey,
+    dk.salt,
+    dk.iv,
+    options,
+  )
+  await fs.ensureDir('keystore')
+  const path = await keythereum.exportToFile(keyObject)
```
Comment (howleysv) [inline packages/toolbox/src/swarm.js]: Seems you can avoid the file copy by providing the path directly and calling 
```
keythereum.exportToFile(keyObject, path)
```

File: packages/toolbox/src/swarm.js (hunk 8 above)
Comment (howleysv) [reply packages/toolbox/src/swarm.js]: According to the docs there
> On Windows, both the forward slash (/) and backward slash (\\) are accepted as path segment separators; however, the path methods only add backward slashes (\\).

Comment (PaulLeCam) [reply packages/toolbox/src/swarm.js]: Better use `path.sep` then: https://nodejs.org/dist/latest-v10.x/docs/api/path.html#path_path_sep

File: packages/toolbox/src/swarm.js (hunk 1 above)
Comment (camronlevanger) [reply packages/toolbox/src/swarm.js]: file is no longer exported here

File: packages/toolbox/src/swarm.js (hunk 2 above)
Comment (camronlevanger) [reply packages/toolbox/src/swarm.js]: done

File: packages/toolbox/src/swarm.js (hunk 5 above)
Comment (camronlevanger) [reply packages/toolbox/src/swarm.js]: removed

File: package.json (hunk 7 above)
Comment (camronlevanger) [reply package.json]: reverted, remnants of exploring, sorry

File: packages/toolbox/package.json [hunk 10]
```diff
@@ -13,7 +13,9 @@
   "dependencies": {
     "@mainframe/config": "*",
     "execa": "^1.0.0",
-    "fs-extra": "^7.0.0"
+    "fs-extra": "^7.0.0",
+    "keytar": "^4.3.0",
```
Comment (PaulLeCam) [inline packages/toolbox/package.json]: I think `keytar` should be added as a dependency to the launcher, not this package.
The main reason is that it requires compiling native code, and this is done by `electron-builder` in the launcher, but if the `keytar` dependency is in this package, it won't be rebuilt.

Comment (camronlevanger) [reply packages/toolbox/package.json]: can we discuss this one further on Monday morning?

Comment (PaulLeCam) [reply packages/toolbox/package.json]: Yes let me know when you're available to chat if you need

File: packages/toolbox/src/swarm.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    '--verbosity',
+    '4',
+    '--bootnodes',
+    'enode://ee9a5a571ea6c8a59f9a8bb2c569c865e922b41c91d09b942e8c1d4dd2e1725bd2c26149da14de1f6321a2c6fdf1e07c503c3e093fb61696daebf74d6acd916b@54.186.219.160:30399',
+    '--ws',
+    '--wsorigins',
+    '*',
+    '--ens-api',
+    'https://mainnet.infura.io/55HkPWVAJQjGH4ucvfW9',
+  ])
```
Comment (howleysv) [inline packages/toolbox/src/swarm.js]: I left comments on this a while ago but they seem to have gotten lost 😓 

Something along the lines of:
- What is the custom bootnode?
- Maybe we should properly set up the CORS config for http and ws
- Is this an old or new infura API key? I believe they're deprecating the old keys
- We should consider using a smaller swarm store than default so that it will calm down syncing sooner. I've been using 100K chunks on my local machine which should be <500MB. Is this big/small enough? 🤷‍♀️

**Comment by PaulLeCam** [comment]
@camronlevanger any update on this please? Should we close it or are you still working on this branch?

**Comment by camronlevanger** [comment]
> @camronlevanger any update on this please? Should we close it or are you still working on this branch?

I'll close. New branch and PR for next release.
